packages/virtual/src/index.tsx

```diff
@@ -1,71 +1,120 @@
+ * A headless virtualized list (see https://www.patterns.dev/vanilla/virtual-lists/) utility for constructing your own virtualized list components with maximum flexibility.
  *
- * @param children the flowComponent that will be used to transform the items into rows in the list
- * @param class the class applied to the root element of the virtualizedList
- * @param each the list of items
- * @param overscanCount the number of elements to render both before and after the visible section of the list, so passing 5 will render 5 items before the list, and 5 items after. Defaults to 1, cannot be set to zero. This is necessary to hide the blank space around list items when scrolling
- * @param rootHeight the height of the root element of the virtualizedList itself
+ * @param items the list of items
+ * @param rootHeight the height of the root element of the virtualizedList
  * @param rowHeight the height of individual rows in the virtualizedList
- * @return virtualized list component
+ * @param overscanCount the number of elements to render both before and after the visible section of the list, so passing 5 will render 5 items before the list, and 5 items after. Defaults to 1, cannot be set to zero. This is necessary to hide the blank space around list items when scrolling
+ * @returns {VirtualListReturn} to use in the list's jsx
  */
-export function VirtualList<T extends readonly any[], U extends JSX.Element>(props: {
-  children: (item: T[number], index: Accessor<number>) => U;
-  fallback?: JSX.Element;
-  class?: string;
-  each: T | undefined | null | false;
-  overscanCount?: number;
-  rootHeight: number;
-  rowHeight: number;
-}): JSX.Element {
-  let rootElement!: HTMLDivElement;
+export function createVirtualList<T extends readonly any[]>({
+  items,
+  rootHeight,
+  rowHeight,
+  overscanCount,
+}: VirtualListConfig<T>): VirtualListReturn<T> {
+  items = access(items) || ([] as any as T);
+  rootHeight = access(rootHeight);
+  rowHeight = access(rowHeight);
+  overscanCount = access(overscanCount) || 1;
 
   const [offset, setOffset] = createSignal(0);
-  const items = () => props.each || ([] as any as T);
 
-  const getFirstIdx = () =>
-    Math.max(0, Math.floor(offset() / props.rowHeight) - (props.overscanCount || 1));
+  const getFirstIdx = () => Math.max(0, Math.floor(offset() / rowHeight) - overscanCount);
 
   const getLastIdx = () =>
     Math.min(
-      items().length,
-      Math.floor(offset() / props.rowHeight) +
-        Math.ceil(props.rootHeight / props.rowHeight) +
-        (props.overscanCount || 1),
+      items.length,
+      Math.floor(offset() / rowHeight) + Math.ceil(rootHeight / rowHeight) + overscanCount,
     );
 
+  return [
+    () => ({
+      containerHeight: items.length * rowHeight,
+      viewerTop: getFirstIdx() * rowHeight,
+      visibleItems: items.slice(getFirstIdx(), getLastIdx()) as unknown as T,
+    }),
+    e => {
+      // @ts-expect-error
+      setOffset(e.target?.scrollTop);
```

Comment:
even if you used `?` it will assign `undefined` to as the offset
Which will either cause an unobvious error later, or just be interpreted as `0`
So better to be explicit about that the behavior should be.

set to zero
```suggestion
      setOffset(e.target ? e.target.scrollTop : 0);
```

or keep the old value
```suggestion
      if (e.target) setOffset(e.target.scrollTop);
```

Reply:
@thetarnav the typescript error being silenced is `Property 'scrollTop' does not exist on type 'EventTarget'.ts(2339)` so that'll still require an `// @ts-expect-error`, but I made the change

Reply:
ah makes sense
that should be fine